### DISCUSSION TITLE
the impact of multiple edges

### DISCUSSION THREAD

**Discussion by xyqbest** [pyg-team/pytorch_geometric#4775] [Q&A]
:raised_hand: Background：
I run a GNN on a graph with more than 350 nodes. Each graph has 64 historical frames. The length of `edge_index` in each frame is different (approximately 8 000 ~ 20 000 edges ). The output is node value in the next frame.
My GNN can deal with each frame one by one under `batch_size=1`.  It seems that there are 64 batches for GNN model.
If my GNN deal with `batch_size=40`, then the input will `torch.cat` the k-th frame over all 40 batches. As a result, the `edge_index` will very long, almost 512 000 ~ 1 280 000 edges. They are lots of repeated edges. It seems that there are 40*64 batches for GNN model.

:question: My question
1. Should I `torch.unique` them before I put them into GNN? I mean removing all repeated `edge_index`.
2. What is the difference between multiple edges and simplified edges? I also run a test, code is in below. I found the output is different. It means that I should not remove multiple edges?

```
import os
import random

import numpy as np
import torch
from torch_geometric.nn import GCNConv


def seed_everything(seed=666):
    # for reproducibility
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True


if __name__ == '__main__':
    seed_everything(seed=666)
    x = torch.rand(10, 7)

    tmp = torch.randint(0, 10, (30, 2))
    # tmp = torch.cat([tmp, tmp[:8]]) # comment this line in the fist run. uncomment this line in the second run to create multiple edges.

    edge_index = torch.LongTensor(tmp.T)
    tmp_layer = GCNConv(7, 4)

    out = tmp_layer(x, edge_index)
    print(out)

```
Thank you :smile:

**Comment by rusty1s** [comment]
How do you encode the different edges across multiple frames? I don't think there is any opportunity to call `torch.unique()` here since values in `edge_index` are typically not duplicated (or at least they shouldn't be): Each frame should be saved in its own subgraph. If you are running OOM when processing 64 frames across 40 graphs (`batch_size=40`), why not simply reduce the batch size?

If you have duplicated edges, that means that you will receive multiple messages coming from the same neighbor. As such, you give a "higher weight" to this kind of neighbor.

**Reply by xyqbest** [reply]
>If you have duplicated edges, that means that you will receive multiple messages coming from the same neighbor. As such, you give a "higher weight" to this kind of neighbor. 

This is what I want.

By the way,
When I use the loader with batch size. The data will `torch.cat` `edge_index` . For example, 
`[Data(..., edge_index=[2,20],...),Data(..., edge_index=[2,22],...),Data(..., edge_index=[2,25],...)]`
with the `batch_size = 3` , the loader will load data `Data(..., edge_index=[2,67],...)` and put it into the model. There are some duplicated edges among `edge_index=[2,67]`. 

Does it impact the output of model? make some neighbors "higher weight"?
Does it lead to the different results compared to `batch_size=1`?

**Reply by rusty1s** [reply]
For mini-batching, there will not be any overlap between edges across different examples. This is described [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html). Importantly, batching will not result in different results compared to `batch_size=1`.
